packages/system/src/Stream/Stream/effectAsyncInterrupt.ts

```diff
@@ -17,12 +21,79 @@ import { effectAsyncInterruptEither } from "./effectAsyncInterruptEither"
  */
 export function effectAsyncInterrupt<R, E, A>(
   register: (
-    cb: (
-      next: T.Effect<R, Option.Option<E>, A.Chunk<A>>,
-      offerCb?: F.Callback<never, boolean>
-    ) => T.UIO<Ex.Exit<never, boolean>>
-  ) => T.Canceler<R>,
+    emit: Emit<R, E, A, Promise<boolean>>
+  ) => E.Either<Canceler<R>, Stream<R, E, A>>,
   outputBuffer = 16
 ): Stream<R, E, A> {
-  return effectAsyncInterruptEither((cb) => E.left(register(cb)), outputBuffer)
+  return new Stream(
+    pipe(
+      M.do,
+      M.bind("output", () =>
+        pipe(
+          Q.makeBounded<Take.Take<E, A>>(outputBuffer),
+          T.toManagedRelease(Q.shutdown)
+        )
+      ),
+      M.bind("runtime", () => pipe(T.runtime<R>(), T.toManaged)),
+      M.bind("eitherStream", ({ output, runtime }) =>
+        M.succeedWith(() =>
+          register(
+            toEmit(async (k) => {
+              try {
+                return runtime.runPromise(
+                  pipe(
+                    Take.fromPull(k),
+                    T.chain((_) => Q.offer_(output, _))
+                  )
+                )
+              } catch (e) {
+                if (e instanceof CS.FiberFailure) {
```

Comment:
better to avoid instanceof comparism

Reply:
also is there a specific reason to use a promise? runPromise hides the full cause